setup_test.go

```diff
@@ -258,6 +258,122 @@ func TestTeeCore(t *testing.T) {
+		if err != nil {
+			t.Errorf("GetLogLevel() returned error: %v", err)
+		} else if lvl != zapcore.Level(expected).String() {
+			t.Errorf("GetLogLevel() = %v, want %v", lvl, zapcore.Level(expected).String())
+		}
+
+		// explicit "*"
+		lvl, err = GetLogLevel("*")
+		if err != nil {
+			t.Errorf(`GetLogLevel("*") returned error: %v`, err)
+		} else if lvl != zapcore.Level(expected).String() {
+			t.Errorf(`GetLogLevel("*") = %v, want %v`, lvl, zapcore.Level(expected).String())
+		}
+
+		// empty string
+		lvl, err = GetLogLevel("")
+		if err != nil {
+			t.Errorf(`GetLogLevel("") returned error: %v`, err)
+		} else if lvl != zapcore.Level(expected).String() {
+			t.Errorf(`GetLogLevel("") = %v, want %v`, lvl, zapcore.Level(expected).String())
+		}
+	}
+}
+
+func TestGetAllLogLevels(t *testing.T) {
+	originalConfig := GetConfig()
+	defer SetupLogging(originalConfig)
+
+	// Clear any state from previous tests first
+	clearLoggerState()
+
+	SetupLogging(Config{Level: LevelWarn, Stderr: true})
+	base := GetAllLogLevels()
+
+	if len(base) != 1 {
+		t.Errorf("baseline GetAllLogLevels() length = %d; want 1", len(base))
+	}
+	if base["*"] != zapcore.Level(LevelWarn).String() {
+		t.Errorf("baseline GetAllLogLevels()[\"*\"] = %v; want %v", base["*"], zapcore.Level(LevelWarn).String())
+	}
+
+	expected := map[string]LogLevel{
+		"test1": LevelDebug,
+		"test2": LevelInfo,
+		"test3": LevelWarn,
+	}
+	SetupLogging(Config{
+		Level:           LevelError,
+		SubsystemLevels: expected,
+		Stderr:          true,
+	})
+
+	all := GetAllLogLevels()
+
+	if all["*"] != zapcore.Level(LevelError).String() {
+		t.Errorf(`GetAllLogLevels()["*"] = %v; want %v`, all["*"], zapcore.Level(LevelError).String())
+	}
+	for name, want := range expected {
+		got, ok := all[name]
+		if !ok {
+			t.Errorf("missing key %q in GetAllLogLevels()", name)
+			continue
+		}
+		if got != zapcore.Level(want).String() {
+			t.Errorf(`GetAllLogLevels()["%s"] = %v; want %v`, name, got, zapcore.Level(want).String())
+		}
+	}
+
+	// dynamic logger test
+	_ = Logger("dynamic")
+	if err := SetLogLevel("dynamic", "fatal"); err != nil {
+		t.Fatalf("SetLogLevel(dynamic) failed: %v", err)
+	}
+
+	all = GetAllLogLevels()
+	if lvl, ok := all["dynamic"]; !ok {
+		t.Error(`missing "dynamic" key after creation`)
+	} else if lvl != zapcore.Level(LevelFatal).String() {
+		t.Errorf(`GetAllLogLevels()["dynamic"] = %v; want %v`, lvl, zapcore.Level(LevelFatal).String())
+	}
+
+	// ensure immutability
+	snapshot := GetAllLogLevels()
+	snapshot["*"] = zapcore.Level(LevelDebug).String()
+	snapshot["newkey"] = zapcore.Level(LevelInfo).String()
+
+	// ensure original state unchanged
+	fresh := GetAllLogLevels()
+	if fresh["*"] != zapcore.Level(LevelError).String() {
+		t.Errorf(`immutable check failed: fresh["*"] = %v; want %v`, fresh["*"], zapcore.Level(LevelError).String())
+	}
+	if _, exists := fresh["newkey"]; exists {
+		t.Error(`immutable check failed: "newkey" should not leak into real map`)
+	}
+}
```

Comment:
If you move the new functions into `levels.go` then the new tests should move to `log_level_test.go`

Reply:
fixed with [`174cfdb` (#160)](https://github.com/ipfs/go-log/pull/160/commits/174cfdbd11a553230949ce958e66a90f07772ab6)